### PR TITLE
Support java.util.StringJoiner

### LINKED ISSUE #3355: Support java.util.StringJoiner
refs: https://github.com/scala-js/scala-js/issues/4872

### PR DESCRIPTION
**Pull request by spamegg1**
- ported from Scala.js https://github.com/scala-js/scala-js/pull/4873
- fixes #3355

### CODE REVIEW

**Comment by LeeTibbert** [comment]
@spamegg1 

Thank you for submitting this PR.

The "lint" failure can be corrected by running scalafmt at the command line.  At the project root "scripts/scalafmt",
then git add, git commit, git push.  Yell, if that does not work for you.    I usually wait for the rest of CI to finish
to see if there are other errors which can be corrected.

**Comment by LeeTibbert** [comment]
Almost there, the failures are because `Platform.hasCompliantNullPointers` is not defined. But then
we knew that.

The third alternative in one of my review comments suggests defining `hasCompliantNullPointers` 
with appropriate comments, at the top of the file and not mucking around with Platform.

File: unit-tests/shared/src/test/scala/org/scalanative/testsuite/javalib/util/StringJoinerTest.scala
Comment (spamegg1) [reply unit-tests/shared/src/test/scala/org/scalanative/testsuite/javalib/util/StringJoinerTest.scala:136]: My original approach was to just remove it entirely. But commenting sounds good.

File: unit-tests/shared/src/test/scala/org/scalanative/testsuite/javalib/util/StringJoinerTest.scala
Comment (LeeTibbert) [inline unit-tests/shared/src/test/scala/org/scalanative/testsuite/javalib/util/StringJoinerTest.scala:168]: Nice handling of complexity.

A native speaker of English would probably use "have" instead of "has".  Not worth a CI run, leave some room for improvement for the next generation.

The content is what is important, IMO.

**Comment by LeeTibbert** [comment]
Looks good to me (LGTM). Well done @spamegg1

File: javalib/src/main/scala/java/util/StringJoiner.scala
Comment (WojciechMazur) [inline javalib/src/main/scala/java/util/StringJoiner.scala:48]: Scala.js has used string concat here becouse string are native construct in JS runtime, and this action does not contain performance penalty. However, on JVM or Native we don't have native String primitives and their concatontation is expensive. We should use `java.lang.StringBuilder` instead.

**Review by WojciechMazur — CHANGES_REQUESTED** [review]

File: javalib/src/main/scala/java/util/StringJoiner.scala
Comment (spamegg1) [reply javalib/src/main/scala/java/util/StringJoiner.scala:48]: Something like this?
```scala
    val builder = java.lang.StringBuilder(value)
    if (isEmpty)
      isEmpty = false
    else
      builder.append(delimiter)

    if (newElement == null)
      builder.append("null")
    else
      builder.append(newElement)

    value = builder.toString()
    this
```

Comment (WojciechMazur) [reply javalib/src/main/scala/java/util/StringJoiner.scala:48]: yeah, something like that. Should do the job. We probably don't need to store result in value

File: javalib/src/main/scala/java/util/StringJoiner.scala
Comment (spamegg1) [reply javalib/src/main/scala/java/util/StringJoiner.scala:51]: Do I still use `+` to add `prefix` and `suffix`?

File: javalib/src/main/scala/java/util/StringJoiner.scala
Comment (WojciechMazur) [inline javalib/src/main/scala/java/util/StringJoiner.scala:20]: It's works nice in JS, but for native/jvm we can do better. The main operation for adding new string is `add(CharSequence)`. Upon String materialization in `toString` we need to add custom prefix and suffix. Instead of constructing intermiediete string we can store parts of it comming from `add`. We can use the fact the List prepend in Scala is very cheap. Here's how I think it could be implemented to more efficient in jvm/native. We have extreamly fast add, and slower materializaiton

```scala
private def delimiter, prefix, suffix: CharSequence 
private def emptyValue: CharSequence | Null 
private var parts: List[CharSequence] = Nil 

def add(v: CharSequence) = { parts ::= v; this } 
def toString() = {
 if parts.isEmpty && emptyValue != null 
 then emptyValue 
 else {
   val sb = new java.lang.StringBuilder(this.length()) // alloc exactly as much buffer as required to skip internal buffer resize
   sb.append(prefix) 
   val head :: tail = parts.reverse(): @unchecked // we know parts is non-empty, parts are in reverse order
   sb.append(head)
   tail.foreach{ part => sb.append(delimiter).append(part) }
   sb.append(suffix)
   sb.toString()
 }
 def merge(other: StringJoiner): StringJoiner = { 
    // take all the parts from other and join 
    if (!other.isEmpty){
      parts = other.parts ::: this.parts
    }
    this
 }
 def length(): Int = {
   val baseLength = if(emptyValue != null) emptyValue.length else prefix.length + suffix.length() 
   val delimLength = delimiter.length() 
   parts.foldLeft(baseLength){case (acc, part) => acc + delimLength + part.length }
 }
}
```

Comment (spamegg1) [reply javalib/src/main/scala/java/util/StringJoiner.scala:20]: Is `parts.reverse()` (followed by `.foreach`) costly here? Could we use `parts.foldRight` maybe? Is that better, or about the same?

Comment (spamegg1) [reply javalib/src/main/scala/java/util/StringJoiner.scala:20]: It looks like, with `merge`, we have to use the delimiter of `other` for parts that come from `other`:
```scala
scala> val multiple = StringJoiner(";", "[", "]").setEmptyValue("--").add("a").add("b").add("c")
val multiple: java.util.StringJoiner = [a;b;c]
                                                                                                                      
scala> val joiner = StringJoiner(", ", "{", "}").add("one").merge(multiple).add("two")
val joiner: java.util.StringJoiner = {one, a;b;c, two}
// here a;b;c is using delimiter of multiple, not of joiner
```
So the `parts = other.parts ::: this.parts` approach doesn't work (or maybe I'm not seeing something obvious). 
What should I do?

Comment (WojciechMazur) [reply javalib/src/main/scala/java/util/StringJoiner.scala:20]: Oh, ok, that's oversight from me. So docs state that the other.parts are concatanate using other delimieter, and it should not add prefix/suffix or empty value. So we can do either do 1 one of things here: 
 1. flatMap other.parts to create a list of delimiters and parts and then prepend them to this.parts 
 2. create new StringBuilder (which is also an instance of CharSequence) and materialize the other joiner without prefix/suffix or empty value. For that we can just copy-past code from toString with minor modifications, or even refector them out to common function. I think this solution might be better.

Comment (spamegg1) [reply javalib/src/main/scala/java/util/StringJoiner.scala:20]: I think I kinda understand what you're saying. 
I'll have to figure out the string length of (`other.parts` + the delimiters that go between them) for the new `StringBuilder`, I might also refactor that into another helper function and reuse it in `length()`.

**Comment by ekrich** [comment]
Just another monkey wrench but you should only use javalibs not Scala but you can use `ScalaOps` to get some Scala like operations. https://github.com/scala-native/scala-native/blob/main/javalib/src/main/scala/java/util/ScalaOps.scala

**Comment by spamegg1** [comment]
> you should only use javalibs not Scala but you can use `ScalaOps` to get some Scala like operations.

I have an implementation that passes the tests. What can I use in place of `scala.collection.immutable.List`? (Hopefully something that has pattern matching?)

**Comment by LeeTibbert** [comment]
FYI, the reason for the "please do not use scalalib" is to break a dependency where scalalib relies upon
javalib which relies upon scalalib.   We are trying to eventually get rid of all scalalib usage in the javalib
code.  If you are reading existing code, you will see many such, mostly from the early days of Scala Native.
The current practice is to be stricter with new code being added or undergoing major renovation.

For me, most of the time knowing the reasoning for a prohibition makes it easier to live with it.

**Comment by spamegg1** [comment]
@LeeTibbert 
> I also usually use a simple `List` and just avoid any Scala specific methods on List. That is use only the methods defined in JVM. One can be a little more explicit and specify `java.util.List`. 

Considering Wojciech's approach above (relying on prepend `::` being cheap), how can I use `java.util.List` to achieve something similar?

>  I did a quick look through the StringJoiner.scala file and nothing jumped out as "non-JVM". 

My current implementation changed completely, I haven't pushed the changes yet. I'm using `scala.collection.immutable.List[CharSequence]`.

**Comment by ekrich** [comment]
@spamegg1 I really should have given more info like Lee has done. I have been doing this awhile so I wasn't really communicating enough information. `ScalaOps` gives you basic Scala Operations like `foreach` and the like so you can use `java.util.List` and `java.util.Set` etc. and not have to give up all the niceties of Scala.

We try to learn from Scala.js which has removed all the Scala stdlib from their javalib so we are headed that direction. Most of our collection code has examples because much of it has been ported from Scala.js so that is a good resource many times. Here is some code, it has 2 annotations from the Scala stdlib but not much else.

https://github.com/scala-native/scala-native/blob/main/javalib/src/main/scala/java/util/Properties.scala

This has examples of `ScalaOps.exists|foldLeft`. https://github.com/scala-native/scala-native/blob/main/javalib/src/main/scala/java/util/IdentityHashMap.scala

File: javalib/src/main/scala/java/util/StringJoiner.scala
Comment (ekrich) [inline javalib/src/main/scala/java/util/StringJoiner.scala]:
```suggestion
// at the top
import java.{util => ju}
import ScalaOps._

  private def lengthOf(
      parts: ju.List[CharSequence],
      delim: CharSequence
  ): Int = {
    val delimLength = delim.length
    if (parts.length = 0) 0
    else {
      val headLen = parts(0).length
      parts.subList(1, parts.length).ScalaOps.foldLeft(headLen)((acc, part) =>
        acc + delimLength + part.length
      )
    }
  }
```
This should be close to what you need. Hope this helps.

**Comment by LeeTibbert** [comment]
@spamegg1 CI is all green, great. 

I want you/us to succeed early, especially on a first PR.  


Without holding up this PR, let me mention a few things for one or more  possible PRs.
I mention these as encouragement and to illustrate some of the factors that go into
SN library code.  Also the power of supportive review to improve the product (but not
at the last minute).

I have heard the expression "Those who propose, dispose." more than 
once. (Meaning, "You think it is a good idea, you do it." Said to keep projects from
ballooning with 1+s and never getting shipped). 

* While reviewing the later change to this code, I discovered that two `String.join()` are not
    implemented. Carve another notch on your coup stick. 
    I created PR #3408 to keep track of the issue.  

* In `StringJoiner.scala` line 33, it seems like `private var isEmpty: Boolean = true` and its
   associated bookkeeping later can be replaced by `contents.isEmpty()` in `toString()` and `merge()`.
   This gets rid of a `var` and bookeeping.  Taking the size of an ArrayList is more expensive than
   accessing a variable, but it is pretty cheap and easy to get right.   I suspect the crossover point
   where one method call is cheaper than a series of increments (in adds) is probably about 10 or so.
   I have no data to backup that WAG (guess).

* The implementation of `StringJoiner.length` looks easy to get right but expensive.  Instantiating the
   String and doing nothing beyond taking its length means doing a, potentially, lot of memory allocations.

   Seems like a follow on could determine the length by doing the math. All the lengths are known at the 
   time of the call. Something like (untried): 
       lengthOfPrefix + lengthOfSuffix + (lengthOfSeparator * lengthOfArrayList) + 
       contents.foldLeft(0)((a: Int, b: String) => a + b.length())

   The potentially painful part is that Scala 2 can have problems with anonymous_functions/lambdas,
   especially with certain version of Java. Scala2.12 & Java 17 has been giving me a kicking several times
   this week.

   I mention this to cut down frustration if a first submission to CI fails.  

   Wojciech has fixed my `StreamTest.scala` several (three or more, set face red) times., so the
   `iterate()` tests there give an example of making a lambda robust.  I usually just create a 
    local `def` for the function and use that.  (I did such two days or so ago. If you want, I can
   try to find the file.)

**Comment by LeeTibbert** [comment]
@spamegg1 
The timing of this PR is excellent, see Issue #3409.  It sure has flushed out a number of bugs.
Thank you

**Comment by LeeTibbert** [comment]
Bug reports are a sign that someone is actually using the product of your hard work (I keep telling myself that 
when I get bug reports ;-)).

I tried using the three argument constructor of `StringJoner` in this PR from `java.util.streamCollectors.scala`.
I got compilation errors, which sent me off to the races of reading both the Scala.js code and this Scala Native code.

1) Both have a `@inline` annotation.  That may be good for Scala.js. I can understand its "this was supposed to be
    an easy, straightforward port' transfer to SN. I suspect/believe that the `@inline` should be deleted. My concern
    is that it will generate lots of code at each use site. (and I am using it in private work).

2) The problem at hand.  The current code declares the StringJoiner constructor with private args.
```
final class StringJoiner  private (                                                   
    delimiter: CharSequence,                                                    
    prefix: CharSequence,                                                       
    suffix: CharSequence                                                        
) extends AnyRef {  
```
I believe this makes the constructor inaccessible (outside the package, including sub-packages.??)

Scala.Js declares the primary constructor as private but then has a 1-arg & 3-arg public method (corresponding
to Java API (edited):
```
final class StringJoiner private (delimiter: String, prefix: String, suffix: String) extends AnyRef {

  def this(delimiter: CharSequence) =
    this(delimiter.toString(), "", "")

  def this(delimiter: CharSequence, prefix: CharSequence, suffix: CharSequence) =
    this(delimiter.toString(), prefix.toString(), suffix.toString())
```
I _think_ that the `toString()` will trigger the NullPointerExceptions described in the API without
an explicit Objects.NotNull() check.

There must be a good reason why the code of this PR is the way it is. Sorry if I am missing it.

**Comment by spamegg1** [comment]
> There must be a good reason why the code of this PR is the way it is. Sorry if I am missing it.

It's my mistake. I wasn't sure but I thought switching to `CharSequence` made the alternate constructor redundant since every `String` is a `CharSequence`. Didn't pay attention to the `private` keyword there! (I've never seen it used that way in Scala code before.)

**Comment by LeeTibbert** [comment]
@spamegg1   I need to apologize.  You may be confused because it looks like I gave you advise which
conflicts with the `CharSequence` advice given above by @WojciechMazur. In re-reading this series of 
replys, I just came across that.  Sorry.

I think the part that is tripping me/us up in trying to convert the original .Js code to SN is the
convenient but limiting Scala `mkstring()`.  That requires String arguments, and the pressure to
use Strings backs up from there all the way to the constructor.

**Comment by LeeTibbert** [comment]
I have spent several hours on this in the middle of the night and have run out of time.
I will be offline most of the time until Thursday.

I created a private study version of StringJoiner.scala.  That revealed a problem with having
the primary constructor arguments as CharSequence.  As the `testState` test in StringJoinerTest 
shows, those arguments can be changed from outside after they are passed to the StringJoiner
argument.  That means that StringJoiner needs to make a deep copy.  Taking a deep copy of a `StringBuilder`
or a `CharSequence` other than `String` is not going to be simple.  Cost/benefit considers may well drive
us to using `String` as the simplest copy.

I suggest that this PR, assuming it passes CI, be merged and remaining issues be addressed in one or
more followup PRs.

I will check in on Thursday to see where the discussion goes.

**Comment by spamegg1** [comment]
@LeeTibbert 
Yes in my previous approach I was making immutable `String` copies of the parameters to fix `testState`.

In the latest commit I changed the primary (private) constructor to use `String` instead, while the auxiliary (public) constructor still uses `CharSequence`. So `testState` is fine now.

**Comment by LeeTibbert** [comment]
This PR gives a good outward facing base level [^1]. The reasonably done is far better than the absent
prefect.

I have a next evolution PR ready to submit after this PR is merged.  It uses a `StringBuilder` and some other
details to reduce the number of String allocation and character copies in `StringJoiner.scala`.  It passes 
`StringJoinerTest` of this PR, without changing the tests.    I have not done performance metrics
but think it will be correct and have a lower resource cost. 

I think submitting the file as a follow-on PR will cause less confusion than posting a gist.
Let's get a firm baseline established.

[^1]  I believe there is one small bug in that `CharSequences` can be altered after they have been passed
         to `add()` and the mutation shows up in `StringJoiner.toString()`.  I have not written a test but
         have inspected the code.

**Review by WojciechMazur — APPROVED** [review]

**Comment by LeeTibbert** [comment]
@spamegg1 

Congratulation on the merge of your first Scala Native PR. 

Programming/Software_Engineering in the large.

Looking forward to your next PR.

**Pull request merged**
